maze-consulting/SFCC_Klaviyo_Maze
# Feature/KSCU-58: Edge cases - bonus products (kscu-58), promotions (kscu-59), discounts (kscu-60)

## Description

This adjustment updates the `dataObj` and `eventData` to accommodate Promotions _([KSCU-59](https://themazegroup.atlassian.net/browse/KSCU-59))_, Discounts _([KSCU-60](https://themazegroup.atlassian.net/browse/KSCU-60))_ and Bonus Products _([KSCU-58](https://themazegroup.atlassian.net/browse/KSCU-58))_.  A need to make these changes was discovered while evaluating each of these, which are subtasks of the edge cases outlined in [KSCU-56](https://themazegroup.atlassian.net/browse/KSCU-56).

With these changes, original price is noted in the tracked data in their own property, promotional & discounted prices are correctly reflected as product prices in key events and bonus products now have a special flag to differentiate bonus products from other products in the _'Started Checkout'_ and _'Order Confirmation'_ events. 


## Manual Testing Steps

This was checked in both SFRA & SiteGen in the following ways as part of manual testing:

- Adding & purchasing a product with a  promotional price (confirm. expected data in returned objs & klaviyo dashboard for each event)
- Adding & purchasing a product with a discounted price (confirm. expected data in returned objs & klaviyo dashboard for each event)
- Adding & purchasing a product with bonus products (confirm. expected data in returned objs & klaviyo dashboard for each event)
- Adding & purchasing a mix of products with & without promos, discounts, bonus products, product options...clothing, electronics with product options, bundles with product options, etc. (confirm. expected data in returned objs & klaviyo dashboard for each event)
- Confirming cart rebuilding links completed for earlier work continue to work as expected for SiteGen & SFRA (ex: successfully creates a link that can be used to recreate a cart...**_which excludes any bonus products_**)

## Pre-Submission Checklist:

[ x ] This update successfully builds
[ x ] No console errors are displayed on the screen with these updates
[ x ] Console Logs & debugging comments created during development that are no longer needed have been removed
[ n/a ] ToDo comments were left as necessary for future reference
[ x ] A description of this update & references to the manual tests has been included in this PR

## Special Reviewer Note:
This branch was created off the [edge case branch for KSCU-65](https://github.com/maze-consulting/SFCC_Klaviyo_Maze/tree/feature/KSCU-65_edge-cases__product-options). The PRs for [product options](https://github.com/maze-consulting/SFCC_Klaviyo_Maze/pull/24) and [product bundles](https://github.com/maze-consulting/SFCC_Klaviyo_Maze/pull/25) should be merged **BEFORE this PR**.
